File: LeetCode/0064. Minimum Path Sum/solution.cpp
/**
 * 61 / 61 test cases passed.
 * Runtime: 4 ms
 * Memory Usage: 9.3 MB 
 */
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        for (int i = 1; i < m; ++ i) grid[i][0] += grid[i - 1][0];
        for (int i = 1; i < n; ++ i) grid[0][i] += grid[0][i - 1];
        for (int i = 1; i < m; ++ i) {
            for (int j = 1; j < n; ++ j) {
                grid[i][j] += min(grid[i - 1][j], grid[i][j - 1]);
            }
        }
        return grid[m - 1][n - 1];
    }
};

File: LeetCode/0064. Minimum Path Sum/solution.py
"""
61 / 61 test cases passed.
Runtime: 40 ms
Memory Usage: 16.4 MB
"""
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        for i in range(1, n): grid[0][i] += grid[0][i - 1]
        for i in range(1, m): grid[i][0] += grid[i - 1][0]
        for i in range(1, m):
            for j in range(1, n):
                grid[i][j] += min(grid[i - 1][j], grid[i][j - 1]) 
        return grid[m - 1][n - 1]
                